docs: how to create a new component (#107)
avengineers/spl-core: docs/getting_started/start_new_component.rst

How to Start Working on a :ref:`Component <what_is_a_component>`
################################################################

To support :ref:`component <what_is_a_component>` development, SPL Core provides features to:

- document the component behavior
- test the component behavior
- generate component report with full traceability between documentation, code and tests

> - **all** component artifacts are configurable (content will be included/excluded based on configuration)

Reviewer: follow the grammatic of the bullet points above and rephrase the sentence, like "configure **all** component artifacts [...]

Reviewer 2: ```suggestion
- configure **all** component artifacts (content will be included/excluded based on configuration)
```

`Sphinx <https://www.sphinx-doc.org/>`_ is used to generate the component report.
See the project `conf.py` file for the configuration details.

Detailed Design
***************

The component behavior is described in restructured text in the ``doc`` directory.

Here are the main features supporting component documentation:

* describe the component behavior in text
* create diagrams using `mermaid <https://mermaid-js.github.io/mermaid/>`_
* include/exclude content based on configuration 
* use `sphinx-needs <https://www.sphinx-needs.com/>`_ to create traceability between documentation, code and tests
    

**Traceability**

In order to trace the documentation to the code or tests, you need to add a `sphinx-needs <https://www.sphinx-needs.com/>`_ element.

.. code-block:: rst

    .. spec:: Feature X
        :id: SWDD_COMP-001

        The component shall implement feature X.

> The ``spec`` is used to tag the component specification. It shall have:

Reviewer: The ``spec`` keyword [...]

* a title to very briefly describe the specification. This will be shown in traceability tables.

> * an id to uniquely identify the specification. The id naming convention is ``SWDD_<COMPONENT_NAME>-<NUMBER>``

Reviewer: Give hint, how one can determine the next "free ID".

Author: We should have a dedicated chapter where the component report feature is explained. There we can describe the traceability and how to keep the IDs unique and have a numbering scheme.

> * free text to describe the specification in more details

Reviewer: add "." at the end of the sentence.

**Template**

Here is a ``index.rst`` template for getting you started documenting a component:

.. code-block:: rst
    :linenos:

    Software Detailed Design
    ########################

    Introduction
    ************

    This component is responsible for ...

    Component Description
    *********************

    .. spec:: Feature X
        :id: SWDD_COMP-001

        The component shall implement feature X.

    Internal Behavior
    *****************

    .. spec::  State Machine
        :id: SWDD_COMP-003

        The component main method is implemented as a state machine. The state machine is shown below.

    .. mermaid::

        stateDiagram-v2
            [*] --> STATE_OFF: Initial State
            STATE_OFF --> STATE_ON : Power State != OFF
            STATE_ON --> STATE_OFF : Power State == OFF
    {% if config.FEATURE_Y %}
            STATE_ON --> STATE_Y : Y Started
            STATE_Y --> STATE_ON : Y Stopped
            STATE_Y --> STATE_OFF : Power State == OFF
    {% endif %}

Test Cases
**********

The component tests are written in GTest and are located in the ``test`` directory.

Here are the main features supporting component tests:

> * test cases are written with `GoogleTest <https://google.github.io/googletest/>`_
> * component external interfaces are automatically mocked using Google Mock
> * include/exclude content based on configuration 
> * use `sphinx-needs <https://www.sphinx-needs.com/>`_ to create traceability between documentation, code and tests

Reviewer 2: ```suggestion
* Test cases are written with `GoogleTest <https://google.github.io/googletest/>`_.
* Component external interfaces are automatically mocked using Google Mock.
* Include/exclude content based on configuration.
* Use `sphinx-needs <https://www.sphinx-needs.com/>`_ to create traceability between documentation, code and tests.
```

> 

Reviewer: for all the bullet point above: add "."

**Traceability**

In order to trace the test case to the design, you need to add a `sphinx-needs <https://www.sphinx-needs.com/>`_ element.

.. code-block:: C++

    /*!
    * @rst
    *
    * .. test:: MyComp.testCorrectBehavior
    *    :id: TS_COMP-001
    *    :tests: SWDD_COMP-001
    *
    * @endrst
    */
    TEST(MyComp, testCorrectBehavior)
    {
        CREATE_MOCK(mymock);
        // test code
    }

The `test` is used to tag the component test case. It shall have:

> * a title required to link the test case to the test result. ❗The title must match exactly the test case name.
> * an ``id`` to uniquely identify the test case. The id naming convention is ``TS_<COMPONENT_NAME>-<NUMBER>``
> * a ``tests`` tag to link the test case to the specification. One can link multiple specifications by separating them with a comma.
> * free text to describe the test case in more details

Reviewer 2: ```suggestion
* a title required to link the test case to the test result. ❗The title must match exactly the test case name.
* an ``id`` to uniquely identify the test case. The id naming convention is ``TS_<COMPONENT_NAME>-<NUMBER>``
* a ``tests`` tag to link the test case to the specification. One can link multiple specifications by separating them with a comma.
* free text to describe the test case in more details.
```

> 

Reviewer: for all the bullet point above: add "."

**Template**

Here is a ``test_comp.cc`` template for getting you started testing a component:

.. code-block:: C++
    
    /**
    * @file
    */
    #include <gtest/gtest.h>
    using namespace testing;

    extern "C" {
    #include "comp.h"
    #include "autoconf.h"
    }

    // Auto-generated mockups for this component
    #include "mockup_src_comp.h"

    /*!
    * @rst
    *
    * .. test:: MyComp.testCorrectBehavior
    *    :id: TS_COMP-001
    *    :tests: SWDD_COMP-001
    *
    * @endrst
    */
    TEST(MyComp, testCorrectBehavior)
    {
        CREATE_MOCK(mymock);
        // Make the input interface return 5
        EXPECT_CALL(mymock, ReadExternalVar()).WillOnce(Return(5));
        // Expect the output interface to be called with 10
        EXPECT_CALL(mymock, WriteOutputVar(10)).Times(1);
        MyCompMain();
    }

Implementing the Component
**************************

The component implementation is located in the ``src`` directory.

Here are the main features supporting component implementation:

* use `sphinx-needs <https://www.sphinx-needs.com/>`_ to create traceability between documentation, code and tests
* include/exclude content based on configuration

> 

Reviewer: for all the bullet point above: add "."

**Traceability**

In order to trace the documentation to the code or tests, you need to add a `sphinx-needs <https://www.sphinx-needs.com/>`_ element.

.. code-block:: C

    /**
    * @rst
    * .. impl:: Some function
    *    :id: SWIMPL_COMP-001
    *    :implements: SWDD_COMP-001
    * @endrst
    */
    unsigned int someFunction(int a)
    {
        ...
    }

The ``impl`` is used to tag the component implementation. It shall have:

* a title to very briefly describe the specification. This will be shown in traceability tables.
* an id to uniquely identify the specification. The id naming convention is ``SWIMPL<_<COMPONENT_NAME>-<NUMBER>``
* an ``implements`` tag to link the implementation to the specification. One can link multiple specifications by separating them with a comma.
* free text to describe the specification in more details

> 

Reviewer: for all the bullet point above: add "."
